README.md
# Cordova Plugin for Editing Images / Pictures #

- Android (Java): Crop, Rotate, BlackWhite, Greyscale, Sharpness | Input: file and base64
- iOS (Swift): Crop, Rotate (Draw Shape, NOT EXECUTED as of 11 05 2020), BlackWhite, Greyscale, Sharpness, Brightness | Input: file only
- Quite nice UI, Colors and Icons
- all Descriptions/Labels are in German
- **!! under active Development !!**
- **!! Currently the Callback Data is _NOT_ consistent between iOS and Android, im working on it !!**#

*DISCLAIMER: This plugin was  created for editing Scans/Photos of Documents, not regular Photos. It works anyways, but some functions, like the BlackWhite, were adjusted to work good with Documents and not a Photo of a nice sunset or whatever...*

## Installation ##

    cordova plugin add https://github.com/bendspoons/cordova-plugin-image-edit.git

## Usage ##

    imageEdit.edit(function(success) {
		// success.code = 201
		// success.result = "SUCCESS"
		// success.path = Path to created/edited Image
    }, function(error) {
    	// error.code = INT
		// error.result = STRING
    }, {
		'sourceData'       : 'file:///path/to/image.jpg',
		'sourceType'       : 'file',
		'destType'         : 'png', 	// OPTIONAL - jpg/png, default jpg
		'allowCrop'        : 1, 		// OPTIONAL - 1 or 0, default 1
		'allowRotate'      : 1, 		// OPTIONAL - 1 or 0, default 1
		'allowFilter'      : 1  		// OPTIONAL - 1 or 0, default 1
    });


Success Code:

    - 201 / SUCCESS

Possible Error Codes:

    - 101 / ABORTED
    - 401 / ERROR_OPTION_INPUTDATA_NOTEXISTS
    - 501 / ERROR_OPTION_INPUTTYPE
    - 502 / ERROR_OPTION_INPUTDATA_EMPTY
    - 503 / ERROR_OPTION_INPUTDATA_BASE64

Returns fully qualified image path like file:///saved/to/path/filename.jpg

base64 mode is not recommended, its likely the app will crash unless you hand over a really small image on Adnroid (prbably remove this whole base64 chunk because i do not need it anymore, and you can always save the base64 data to file and use this plugin with that saved file... amiright?!)

*--> Conclusion: i will definitly remove the buggy base64 functionality*

## Edit options ##

### Filter ###

Black / White (on iOS with a little analysing of the given image brightness and according actions)

Greyscale

Sharpness

(Brightness iOS only)
 
### Cropping ###

with Rectangle Drawing
 
### Rotate ###

in 90° steps (on iOS with additinal UISlider)


## Example with Camera ##

    navigator.camera.getPicture(function(cameraPicturePath) {
      	console.log('Camera Picture Path: ' + success.path);

      	imageEdit.edit(function(editedImagePath) {
     		console.log('Edited Image Path: ' + editedImagePath);   
      	}, function(error) {
     		alert('Failed because: ' + JSON.stringify(error);
      	}, {
     		sourceData: cameraPicturePath,
     		sourceType: 'file'
      	});

    }, function(cameraError) {
      alert('Camera failed because: ' + cameraError);
    }, {
      quality: 80,
      destinationType: Camera.DestinationType.FILE_URI
    });


Changelog.md
# Changelog #

## 11-05-2020 ## 
- iOS (Swift) Version
- works with iOS 11 and up, with iOS < 13 the Buttons are labelled with Text, starting with iOS 13 the buttons are styled with SF Icons (systemName:)

## 09-05-2020 ## 
- Android (Java) Version

## 06-05-2020 ##
- Initialization of plugin

example/js/index.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var app = {
    // Application Constructor
    initialize: function() {
        document.addEventListener('deviceready', this.onDeviceReady.bind(this), false);
    },

    // deviceready Event Handler
    //
    // Bind any cordova events here. Common events are:
    // 'pause', 'resume', etc.
    onDeviceReady: function() {
        /*
        setTimeout(function() {
            //var success =  cordova.file.applicationStorageDirectory + 'img/beleg.jpeg';
            var success =  'img/beleg.jpeg';
           console.log('load image from www' + success)
           $('#cameraImage').attr('src', success);
           $('#cameraImage').attr('alt', success);
           $('#cameraImage').attr('title', success);
           $('#cameraImageSrc').html(success);
           $('.show-when-image').show();
        }, 3000);
        */
    },

    camera: function() {
      navigator.camera.getPicture(function(success) {
        console.log('Camera Source: ' + success);

        $('#cameraImage').attr('src', success);
        $('#cameraImageSrc').html(success);
        $('.show-when-image').show();
      }, function(message) {
         alert('Failed because: ' + message);
      }, {
         quality: 80,
         destinationType: Camera.DestinationType.FILE_URI
      });
    },

    editImage: function() {
      imageEdit.edit(function(success) {
        console.log('Edited Image Source: ', JSON.stringify(success));
        alert('Edited Image Source: ' + JSON.stringify(success));

         $('#editedImage').attr('src', success.path);
         $('#editedImageSrc').html(JSON.stringify(success));
         $('.show-when-edited').show();
      }, function(error) {
        console.log('Failed because: ', JSON.stringify(error));
        alert('Failed because: ' + JSON.stringify(error));
      }, {
         'sourceData'       : $('#cameraImage').attr('src'),
         'sourceType'       : 'file',
         //'destType'         : 'png', // jpg or png, default jpg
         //'allowCrop'        : 1, // 1 or 0, default 1
         //'allowRotate'      : 1, // 1 or 0, default 1
         //'allowFilter'      : 1 // 1 or 0, default 1
      });
    },

    setBasicImage: function() {
      var success = 'file:///storage/emulated/0/Android/data/com.bendspoons.imageedit/cache/1588864986783.jpg';
       console.log('Edited Image Source: ' + success);
       $('#cameraImage').attr('src', success);
       $('#cameraImageSrc').html(success);
       $('.show-when-image').show();
    },

    setPickImage: function() {
        window.imagePicker.getPictures(function(results) {

            var imgSrc = 'file://' + results[0];
            console.log('Edited Image Source: ' + imgSrc);
            $('#cameraImage').attr('src', imgSrc);
            $('#cameraImageSrc').html(imgSrc);
            $('.show-when-image').show();

            for (var i = 0; i < results.length; i++) {
                console.log('Image URI: ' + results[i]);
            }
        }, function (error) {
            console.log('Error: ' + error);
        });
    },

    photoAndEdit: function() {
      navigator.camera.getPicture(function(cameraPicturePath) {
        console.log('Camera Picture Path: ' + success);

        imageEdit.edit(function(editedImagePath) {
           console.log('Edited Image Path: ' + editedImagePath);

        }, function(error) {
           alert('Failed because: ' + error);
        }, {
           sourceData: cameraPicturePath,
           sourceType: 'file'
        });
      }, function(cameraError) {
         alert('Camera failed because: ' + cameraError);
      }, {
         quality: 80,
         destinationType: Camera.DestinationType.FILE_URI
      });
    }
};

app.initialize();


src/android/ImageEdit.java
package com.bendspoons.plugin.imageedit;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Base64;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.LOG;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.UUID;

import com.bendspoons.plugin.imageedit.ImageEditActivity;

/**
 * Created by jt on 29/03/16.
 */
public class ImageEdit extends CordovaPlugin {
    private static final String TAG = "ImageEdit";

    private String sourceData;
    private String sourceType;

    private CallbackContext callbackContext;

    @Override
    public boolean execute(String action, CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        if (!action.equals("edit")) {
            callbackContext.sendPluginResult(
                    new PluginResult(Status.INVALID_ACTION, "Unsupported action: " + action));
            return false;
        }

        try {
            this.sourceData = args.getString(0);
            LOG.e(TAG, "this.sourceData = " + this.sourceData);

            this.sourceType = args.getString(1);
            LOG.e(TAG, "this.sourceType = " + this.sourceType);

            if (this.cordova != null) {
                LOG.e(TAG, "doEdit()");
                doEdit();
            }

            this.callbackContext = callbackContext;
            return true;
        } catch (JSONException e) {
            callbackContext.sendPluginResult(new PluginResult(Status.JSON_EXCEPTION, e.getMessage()));
            return false;
        }
    }

    private void doEdit() {
        this.cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                final Intent touchDrawIntent = new Intent(ImageEdit.this.cordova.getActivity(), ImageEditActivity.class);

                LOG.e(TAG, "ImageEdit.this.sourceData = " + ImageEdit.this.sourceData);
                LOG.e(TAG, "ImageEdit.this.sourceType = " + ImageEdit.this.sourceType);
                //touchDrawIntent.putExtra("dataSource", ImageEdit.this.sourceData);
                //touchDrawIntent.putExtra("dataType", ImageEdit.this.sourceType);

                if (ImageEdit.this.sourceType.equals("base64")) {
                    LOG.e("doEdit", "base64");
                    touchDrawIntent.putExtra("dataType", ImageEdit.this.sourceType);
                    touchDrawIntent.putExtra("dataSource", ImageEdit.this.sourceData);
                } else if (ImageEdit.this.sourceType.equals("file")) {
                    LOG.e("doEdit", "file");
                    Uri inputUri = Uri.parse(ImageEdit.this.sourceData);
                    String scheme = (inputUri != null && inputUri.getScheme() != null) ? inputUri.getScheme() : "";
                    LOG.e("doEdit", "scheme " + scheme);

                    if (scheme.equals(ContentResolver.SCHEME_CONTENT)) {
                        // Workaround for CB-9548 (https://issues.apache.org/jira/browse/CB-9548)
                        //  The Cordova camera plugin can sometimes return a content URI instead of a file URI
                        //  when the image is selected from the photo gallery.
                        //
                        //  However, the ImageEditActivity can only accept a file URI or a data URI for the
                        //  background image. So, we need to read the background image data and pass it in a
                        //  format which can be handled by the ImageEditActivity.

                        InputStream inStream = null;
                        try {
                            // Write background image to a temporary file and pass it as a file URL because
                            // there is no reliable way to get a file path from a content URI
                            // (http://stackoverflow.com/a/19985374)
                            ContentResolver contentResolver = ImageEdit.this.cordova.getActivity().getContentResolver();
                            inStream = contentResolver.openInputStream(inputUri);

                            if (inStream != null) {
                                File file = new File(ImageEdit.this.cordova.getActivity().getCacheDir(), UUID.randomUUID().toString());
                                FileOutputStream outStream = new FileOutputStream(file);
                                byte[] data = new byte[1024];
                                int bytesRead;

                                while ((bytesRead = inStream.read(data, 0, data.length)) != -1) {
                                    outStream.write(data, 0, bytesRead);
                                }
                                outStream.flush();
                                outStream.close();

                                touchDrawIntent.putExtra("dataType", "file");
                                touchDrawIntent.putExtra("dataSource", "file://" + file.getAbsolutePath());
                                LOG.e("doEdit", "dataSource SCHEME_CONTENT " + "file://" + file.getAbsolutePath());
                            }
                        } catch (IOException e) {
                            String message = "Failed to read image data from " + inputUri;
                            LOG.e(TAG, message);
                            e.printStackTrace();

                            ImageEdit.this.callbackContext.error(message + ": " + e.getLocalizedMessage());
                            return;
                        } finally {
                            if (inStream != null) {
                                try {
                                    inStream.close();
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }
                            }
                        }

                    } else if (scheme.equals(ContentResolver.SCHEME_FILE)) {
                        touchDrawIntent.putExtra("dataType", "file");
                        touchDrawIntent.putExtra("dataSource", ImageEdit.this.sourceData);
                        LOG.e("doEdit", "dataSource SCHEME_FILE " + ImageEdit.this.sourceData);
                    } else {
                        LOG.e("doEdit", "dataSource ELSE " + ImageEdit.this.sourceData);
                        String message = "invalid scheme for inputData: " + ImageEdit.this.sourceData ;
                        File file = new File(ImageEdit.this.sourceData);

                        LOG.d(TAG, message);
                        if (file.exists() && !file.isDirectory()) {
                            touchDrawIntent.putExtra("dataType", "file");
                            touchDrawIntent.putExtra("dataSource else", "file://" + file.getAbsolutePath());
                        } else {
                            ImageEdit.this.callbackContext.error(message);
                            return;
                        }
                    }
                }






                ImageEdit.this.cordova.getActivity().runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        LOG.e(TAG, "run()");
                        ImageEdit.this.cordova.startActivityForResult(ImageEdit.this, touchDrawIntent, 900);
                    }
                });
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, final Intent intent) {
        LOG.e("ActivityResult", ""+resultCode);

        if (resultCode == Activity.RESULT_CANCELED) {
            this.callbackContext.success("");
            return;
        }

        if (resultCode == Activity.RESULT_OK && this.cordova != null) {
            this.cordova.getThreadPool().execute(new Runnable() {
                @Override
                public void run() {
                    saveDrawing(intent);
                }
            });
            return;
        }

        if (resultCode == ImageEditActivity.RESULT_TOUCHDRAW_ERROR) {
            Bundle extras = intent.getExtras();
            String errorMessage = "Failed to generate ImageEdit.";

            if (extras != null) {
                errorMessage += " " + extras.getString(ImageEditActivity.DRAWING_RESULT_ERROR);
            }

            this.callbackContext.error(errorMessage);
        }
    }

    private void saveDrawing(Intent intent) {
        Bundle extras = intent.getExtras();
        byte[] drawingData = null;
        String output = null;

        if (extras != null &&
                extras.containsKey("drawing_result")) {
            drawingData = extras.getByteArray("drawing_result");
        }

        if (drawingData == null || drawingData.length == 0) {
            LOG.e(TAG, "Failed to read sketch result from activity");
            this.callbackContext.error("Failed to read sketch result from activity");
            return;
        }

        try {
            String ext = "";

            ext = "jpeg";
            //ext = "png";


            String fileName = String.format("sketch-%s.%s", UUID.randomUUID(), ext);
            File filePath = new File(this.cordova.getActivity().getCacheDir(), fileName);

            FileOutputStream fos = new FileOutputStream(filePath);
            fos.write(drawingData);
            fos.close();

            // Add the drawing to photo gallery
            String appName = getApplicationLabelOrPackageName(this.cordova.getActivity());
            String mediaStoreUrl = MediaStore.Images.Media.insertImage(this.cordova.getActivity().getContentResolver(),
                    filePath.getAbsolutePath(), fileName,
                    (appName != null && !appName.isEmpty()) ? "Generated by " + appName : "");

            LOG.d(TAG, (mediaStoreUrl != null) ?
                    "Drawing saved to media store: " + mediaStoreUrl :
                    "Failed to save drawing to media store");

            // We need to return the file saved to the cache dir instead of the
            // file in the photo gallery because the Cordova file plugin cannot open content URIs
            output = "file://" + filePath.getAbsolutePath();
            LOG.d(TAG, "Drawing saved to: " + output);
        } catch(Exception e) {
            LOG.e(TAG, "Error generating output from drawing: " + e.getMessage());

            this.callbackContext.error("Failed to generate output from drawing: "
                    + e.getMessage());
            return;
        }

        this.callbackContext.success(output);
    }

    // Based on http://stackoverflow.com/a/16444178
    private String getApplicationLabelOrPackageName(Context context) {
        PackageManager pm = context.getPackageManager();
        ApplicationInfo appInfo = context.getApplicationInfo();

        if (pm == null || appInfo == null) {
            return "";
        }

        try {
            String label = (String) pm.getApplicationLabel(pm.getApplicationInfo(appInfo.packageName, 0));
            if (label != null && !label.isEmpty()) {
                return label;
            }
        } catch (PackageManager.NameNotFoundException e) {
            LOG.w(TAG, "Failed to determine app label");
        }

        return appInfo.packageName;
    }
}


src/ios/ImageEdit.swift
@objc(ImageEdit) class ImageEdit : CDVPlugin, ImageEditDelegate {

    var commandCollbackId = "";

    /*
    let RESULT_CODES: [Int] = [200,400,401,402,403]
    let RESULT_MSGS: [String] = [
        "SUCCESS",
        "ABORT",
        "ERROR",
        "ERROR_OPTION_INPUTTYPE",
        "ERROR_OPTION_INPUTDATA"
    ]

    let RESULT_CODES: KeyValuePairs = [
        200: "SUCCESS",
        400: "ABORT",
        401: "ERROR",
        501: "ERROR_OPTION_INPUTTYPE",
        502: "ERROR_OPTION_INPUTDATA_EMPTY",
        503: "ERROR_OPTION_INPUTDATA_BASE64"
    ] as KeyValuePairs<AnyHashable, NSString>
    */

    // initialize pluin
    override func pluginInitialize() {
        print("ImageEdit :: pluginInitialize")
    }

    /* version */
    @objc(edit:)
    func edit(command: CDVInvokedUrlCommand) {
        print("ImageEdit :: edit")

        DispatchQueue.main.async {
            // Store cordova callback
            self.commandCollbackId = command.callbackId;

            print("options \(command.arguments ?? [])")

            // inputType file
            // on iOS, only file is allowed
            let inputType = command.arguments[1] as? String ?? ""
            if(inputType != "file") {
                self.imageEdited(resultCode: 501, resultString: "ERROR_OPTION_INPUTTYPE", imagePath: "");
                return;
            }

            // FileURL
            // on iOS, only fileURLs are allowed
            var fileURL = command.arguments[0] as? String ?? ""
            if(fileURL == "data:image/jpeg;base64,") {
                fileURL = "";
            }

            // empty fileURL
            if(fileURL.isEmpty) {
                self.imageEdited(resultCode: 502, resultString: "ERROR_OPTION_INPUTDATA_EMPTY", imagePath: "");
                return;
            }

            // base64 fileURL
            if(fileURL.contains("data:image/")) {
                self.imageEdited(resultCode: 503, resultString: "ERROR_OPTION_INPUTDATA_BASE64", imagePath: "");
                return;
            }

            let destType = command.arguments[2] as? String ?? "jpg"
            let allowCrop = command.arguments[3] as? Int ?? 1
            let allowRotate = command.arguments[4] as? Int ?? 1
            let allowFilter = command.arguments[5] as? Int ?? 1


            print("ImageEdit :: fileURL ", fileURL)
            print("ImageEdit :: inputType", inputType)

            print("ImageEdit :: destType", destType)
            print("ImageEdit :: allowCrop", allowCrop)
            print("ImageEdit :: allowRotate", allowRotate)
            print("ImageEdit :: allowFilter", allowFilter)

            // ImageEditViewController
            let detailVC = ImageEditViewController()

            detailVC.filePath = fileURL;

            detailVC.allowCrop = allowCrop;
            detailVC.allowRotate = allowRotate;
            detailVC.allowFilter = allowFilter;

            detailVC.destType = destType;

            detailVC.delegate = self;

            // navigationController
            let navigationController = UINavigationController(rootViewController: detailVC)
            navigationController.modalPresentationStyle = .fullScreen

            self.viewController?.present(navigationController, animated: true)
        }
    }

    func imageEdited(resultCode: NSInteger, resultString: NSString, imagePath: NSString) {
        print("ImageEdit :: imageEdited CALLBACK")
        print("ImageEdit :: imageEdited resultCode", resultCode)
        print("ImageEdit :: imageEdited resultString", resultString)
        print("ImageEdit :: imageEdited imagePath", imagePath)

        var pluginResult = CDVPluginResult(
            status: CDVCommandStatus_ERROR
        )

        if(resultCode == 201) { // Success, imagepath returned
            let successResult: [String: Any] = [
              "code"    : resultCode,
              "result"  : resultString,
              "path"    : imagePath
            ]

            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: successResult
            )
        } else { // Abort or Error
            let errorResult: [String: Any] = [
              "code": resultCode,
              "result": resultString
            ]

            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: errorResult
            )
        }

        self.commandDelegate!.send(
            pluginResult,
            callbackId: self.commandCollbackId
        )
    }

    func parseTuple(from string: String) -> (String, Int)? {

        if let theRange = string.range(of: "/", options: .backwards),
            let i = Int(string[theRange.upperBound...]) {
            return (String(string[...theRange.lowerBound]), i)
        } else {
            return nil
        }
    }

}




 /* version
 @objc(abort:)
   func abort(imagePath: NSString) {
   print("ImageEdit :: abort", imagePath)

 }
*/

    /*
     let toastController: UIAlertController =
        UIAlertController(
          title: "",
          message: fileData,
          preferredStyle: .alert
        )

      self.viewController?.present(
        toastController,
        animated: true,
        completion: nil
      )

      DispatchQueue.main.asyncAfter(deadline: .now() + 3) {
        toastController.dismiss(
          animated: true,
          completion: nil
        )
      }
*/
